Source/WebKit/WPEPlatform/wpe/WPEGestureController.cpp

```diff
@@ -0,0 +1,269 @@
+            break;
+        case WPE_EVENT_TOUCH_MOVE:
+            if (!m_touchDownPoint)
+                break;
+            if (double x, y; wpe_event_get_position(event, &x, &y)) {
+                if (!m_dragReferencePoint && std::hypot(x - m_touchDownPoint->x, y - m_touchDownPoint->y) > dragActivationThresholdPx)
+                    m_dragReferencePoint = m_touchDownPoint;
+                if (m_dragReferencePoint) {
+                    detectedGesture = WPE_GESTURE_DRAG;
+                    m_detectedGesturePoint = m_touchDownPoint;
+                    m_detectedGestureDelta = { x - m_dragReferencePoint->x, y - m_dragReferencePoint->y };
+                    m_dragReferencePoint = Point { x, y };
+                }
+            }
+            break;
+        case WPE_EVENT_TOUCH_UP:
+            if (double x, y; wpe_event_get_position(event, &x, &y) && m_touchDownPoint) {
+                if (m_dragReferencePoint) {
+                    detectedGesture = WPE_GESTURE_DRAG;
+                    m_detectedGesturePoint = m_touchDownPoint;
+                    m_detectedGestureDelta = { x - m_dragReferencePoint->x, y - m_dragReferencePoint->y };
+                } else if (m_touchDownPoint) {
+                    detectedGesture = WPE_GESTURE_TAP;
+                    m_detectedGesturePoint = m_touchDownPoint;
+                    m_detectedGestureDelta = std::nullopt;
+                }
+            }
+            reset();
+            break;
+        default:
+            reset();
+            break;
+        }
+
+        if (detectedGesture == WPE_GESTURE_NONE) {
+            m_detectedGesturePoint = std::nullopt;
+            m_detectedGestureDelta = std::nullopt;
+        }
+        m_detectedGesture = detectedGesture;
+
+        return m_detectedGesture;
+    }
+
+    WPEGesture getDetectedGesture() const { return m_detectedGesture; }
+
+    struct Point {
+        double x;
+        double y;
+    };
+    using Delta = Point;
+
+    std::optional<Point> getDetectedGesturePoint() const { return m_detectedGesturePoint; }
+    std::optional<Delta> getDetectedGestureDelta() const { return m_detectedGestureDelta; }
+
+private:
+    void reset()
+    {
+        m_touchDownPoint = std::nullopt;
+        m_dragReferencePoint = std::nullopt;
+    }
+
+    // FIXME: These ought to be either configurable or derived from system
+    //        properties, such as screen size and pixel density.
+    static constexpr uint32_t dragActivationThresholdPx { 8 };
+
+    std::optional<Point> m_touchDownPoint;
+    std::optional<Point> m_dragReferencePoint;
+    WPEGesture m_detectedGesture = WPE_GESTURE_NONE;
+    std::optional<Point> m_detectedGesturePoint;
+    std::optional<Delta> m_detectedGestureDelta;
+};
+
+} // namespace WPE
+
+
+/**
+ * WPEGestureController:
+ *
+ */
+struct _WPEGestureControllerPrivate {
+    WPE::GestureDetector detector;
+};
+
+WEBKIT_DEFINE_TYPE(WPEGestureController, wpe_gesture_controller, G_TYPE_OBJECT)
```

Comment:
There's no common code in the base class for this, so I think this should be an interface instead of an abstract class. We could add `WPEGestureConstrollerImpl` (as private class) with the default implementation to be used by WPEDisplay by default.